File: typescript-design-pattern/src/into-to-design-patterns-using-java-lang/iterator/book-shelf-iterator.ts
import Book from './book'
import BookShelf from './book-shelf'
import { Iterator } from './iterator'

export default class BookShelfIterator implements Iterator<Book> {
  private readonly _bookShelf: BookShelf
  private _index: number

  constructor(bookShelf: BookShelf) {
    this._bookShelf = bookShelf
    this._index = 0
  }

  hasNext(): boolean {
    return this._index < this._bookShelf.length()
  }

  next(): Book {
    if (!this.hasNext()) {
      throw new Error('no such element')
    }
    const book = this._bookShelf.getBookAt(this._index)
    this._index = this._index + 1
    return book
  }
}

File: c++-design-pattern/src/intro_to_design_patterns_using_java_lang/iterator/BookShelfIterator.cpp
#include "BookShelfIterator.h"

#include <iostream>
#include <stdexcept>
#include <utility>

BookShelfIterator::BookShelfIterator() : books{std::make_shared<std::vector<Book>>()}
{
}

BookShelfIterator::BookShelfIterator(std::shared_ptr<std::vector<Book>> books)
    : books(books)
{
}

BookShelfIterator::~BookShelfIterator() = default;

bool BookShelfIterator::hasNext() const
{
    return index < books->size();
}

Book BookShelfIterator::next()
{
    if (!hasNext())
    {
        throw std::out_of_range("no such element");
    }
    Book book = books->at(index);
    index++;
    return book;
}
